main.go

```diff
@@ -142,19 +142,17 @@ func (s *prometheusServer) Start(ctx context.Context) error {
 		return err
 	}
 
+	subCtx, cancel := context.WithCancel(ctx)
 	go func() {
-		// TODO: change context, so it can be closed when
-		// this function returns an error
-		if err := certWatcher.Start(ctx); err != nil {
+		if err := certWatcher.Start(subCtx); err != nil {
 			setupLog.Error(err, "certificate watcher error")
 		}
+		cancel()
```

Comment:
Use `defer cancel()` at the start of this goroutine. That is the usual way to call `cancel()`.